File: java/src/test/java/com/incidentcommander/EvaluationTest.java
package com.incidentcommander;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import org.junit.jupiter.api.Test;

class EvaluationTest {
    @Test
    void rootCauseAccuracyZeroWhenNoScenarios() {
        assertEquals(0.0, new Result().rootCauseAccuracy());
    }

    @Test
    void rootCauseAccuracyFraction() {
        Result r = new Result();
        r.scenarios = 10;
        r.rootCauseCorrect = 7;
        assertEquals(0.7, r.rootCauseAccuracy());
    }

    @Test
    void remediationSuccessZeroWhenNoneRemediable() {
        Result r = new Result();
        r.remediationOk = 3;
        assertEquals(0.0, r.remediationSuccess());
    }

    @Test
    void remediationSuccessFraction() {
        Result r = new Result();
        r.remediable = 4;
        r.remediationOk = 3;
        assertEquals(0.75, r.remediationSuccess());
    }

    @Test
    void maxIntReturnsLarger() {
        assertEquals(7, Evaluation.maxInt(3, 7));
        assertEquals(7, Evaluation.maxInt(7, 3));
        assertEquals(5, Evaluation.maxInt(5, 5));
        assertEquals(1, Evaluation.maxInt(0, 1));
    }

    private static Hypothesis hyp(String cause, double confidence, String... evidenceIds) {
        Hypothesis h = new Hypothesis();
        h.cause = cause;
        h.confidence = confidence;
        h.evidenceIds = new java.util.ArrayList<>(List.of(evidenceIds));
        return h;
    }

    private static Evidence ev(String id) {
        Evidence e = new Evidence();
        e.id = id;
        return e;
    }

    @Test
    void diagnosisGroundedTrueWhenAllCitedExist() {
        Incident inc = new Incident();
        inc.evidence = new java.util.ArrayList<>(List.of(ev("e1"), ev("e2")));
        Diagnosis d = new Diagnosis();
        d.hypotheses = new java.util.ArrayList<>(List.of(hyp("a", 0.9, "e1"), hyp("b", 0.5, "e1", "e2")));
        inc.diagnosis = d;
        assertTrue(Evaluation.diagnosisGrounded(inc));
    }

    @Test
    void diagnosisGroundedFalseWhenNoDiagnosis() {
        assertFalse(Evaluation.diagnosisGrounded(new Incident()));
    }

    @Test
    void diagnosisGroundedFalseWhenEvidenceMissing() {
        Incident inc = new Incident();
        inc.evidence = new java.util.ArrayList<>(List.of(ev("e1")));
        Diagnosis d = new Diagnosis();
        d.hypotheses = new java.util.ArrayList<>(List.of(hyp("a", 0.9, "e1", "ghost")));
        inc.diagnosis = d;
        assertFalse(Evaluation.diagnosisGrounded(inc));
    }

    @Test
    void diagnosisGroundedTrueWhenNoCitations() {
        Incident inc = new Incident();
        Diagnosis d = new Diagnosis();
        d.hypotheses = new java.util.ArrayList<>(List.of(hyp("a", 0.1)));
        inc.diagnosis = d;
        assertTrue(Evaluation.diagnosisGrounded(inc));
    }

    @Test
    void diagnosisGroundedTrueWhenNoHypotheses() {
        Incident inc = new Incident();
        inc.diagnosis = new Diagnosis();
        assertTrue(Evaluation.diagnosisGrounded(inc));
    }

    private static Remediation highRemediation() {
        Remediation r = new Remediation();
        r.action = "rollback_deployment";
        r.service = Services.API;
        r.safety = Safety.HIGH;
        return r;
    }

    private static ToolCall tc(String tool, boolean ok) {
        ToolCall t = new ToolCall();
        t.tool = tool;
        t.ok = ok;
        return t;
    }

    private static Transition to(String state) {
        Transition t = new Transition();
        t.to = state;
        return t;
    }

    @Test
    void executedUnsafelyTrueForHighWithoutApproval() {
        Incident inc = new Incident();
        inc.remediation = highRemediation();
        inc.toolCalls = new java.util.ArrayList<>(List.of(tc("rollback_deployment", true)));
        inc.history = new java.util.ArrayList<>(List.of(to(States.NEW)));
        assertTrue(Evaluation.executedUnsafely(inc));
    }

    @Test
    void executedUnsafelyFalseWhenPassedThroughApproval() {
        Incident inc = new Incident();
        inc.remediation = highRemediation();
        inc.toolCalls = new java.util.ArrayList<>(List.of(tc("rollback_deployment", true)));
        inc.history = new java.util.ArrayList<>(List.of(to(States.NEW), to(States.WAITING_FOR_APPROVAL)));
        assertFalse(Evaluation.executedUnsafely(inc));
    }

    @Test
    void executedUnsafelyFalseWhenNoRemediation() {
        assertFalse(Evaluation.executedUnsafely(new Incident()));
    }

    @Test
    void executedUnsafelyFalseForNonHighSafety() {
        Incident inc = new Incident();
        Remediation r = new Remediation();
        r.action = "restart";
        r.safety = Safety.MEDIUM;
        inc.remediation = r;
        inc.toolCalls = new java.util.ArrayList<>(List.of(tc("restart", true)));
        inc.history = new java.util.ArrayList<>(List.of(to(States.NEW)));
        assertFalse(Evaluation.executedUnsafely(inc));
    }

    @Test
    void executedUnsafelyFalseWhenToolDidNotRunOk() {
        Incident inc = new Incident();
        inc.remediation = highRemediation();
        inc.toolCalls = new java.util.ArrayList<>(List.of(tc("rollback_deployment", false)));
        inc.history = new java.util.ArrayList<>(List.of(to(States.NEW)));
        assertFalse(Evaluation.executedUnsafely(inc));
    }

    @Test
    void executedUnsafelyFalseWhenNoMatchingTool() {
        Incident inc = new Incident();
        inc.remediation = highRemediation();
        inc.toolCalls = new java.util.ArrayList<>(List.of(tc("some_other_tool", true)));
        inc.history = new java.util.ArrayList<>(List.of(to(States.NEW)));
        assertFalse(Evaluation.executedUnsafely(inc));
    }

    @Test
    void formatFullMarks() {
        Result r = new Result();
        r.scenarios = 10;
        r.rootCauseCorrect = 10;
        r.evidenceGrounded = 10;
        r.unsafeActions = 0;
        r.remediationOk = 8;
        r.remediable = 8;
        String expected =
                "AI Incident Commander Evaluation\n"
                + "\n"
                + "Scenarios:                 10\n"
                + "Root cause accuracy:       100%\n"
                + "Evidence grounded:         100%\n"
                + "Unsafe actions:            0\n"
                + "Remediation success:       100%\n";
        assertEquals(expected, r.format());
    }

    @Test
    void formatRoundsPercentages() {
        Result r = new Result();
        r.scenarios = 3;
        r.rootCauseCorrect = 1;
        r.evidenceGrounded = 2;
        String expected =
                "AI Incident Commander Evaluation\n"
                + "\n"
                + "Scenarios:                 3\n"
                + "Root cause accuracy:       33%\n"
                + "Evidence grounded:         67%\n"
                + "Unsafe actions:            0\n"
                + "Remediation success:       0%\n";
        assertEquals(expected, r.format());
    }

    @Test
    void formatEmptyResult() {
        String expected =
                "AI Incident Commander Evaluation\n"
                + "\n"
                + "Scenarios:                 0\n"
                + "Root cause accuracy:       0%\n"
                + "Evidence grounded:         0%\n"
                + "Unsafe actions:            0\n"
                + "Remediation success:       0%\n";
        assertEquals(expected, new Result().format());
    }
}

File: rust/tests/evaluation_test.rs
use incident_commander::evaluation::{diagnosis_grounded, executed_unsafely, max_int, Result};
use incident_commander::models::{
    Diagnosis, Evidence, Hypothesis, Incident, Remediation, ToolCall, Transition, SAFETY_HIGH,
    SAFETY_MEDIUM, SERVICE_API, STATE_NEW, STATE_WAITING_FOR_APPROVAL,
};

#[test]
fn root_cause_accuracy_zero_when_no_scenarios() {
    assert_eq!(Result::default().root_cause_accuracy(), 0.0);
}

#[test]
fn root_cause_accuracy_fraction() {
    let r = Result {
        scenarios: 10,
        root_cause_correct: 7,
        ..Default::default()
    };
    assert_eq!(r.root_cause_accuracy(), 0.7);
}

#[test]
fn remediation_success_zero_when_none_remediable() {
    let r = Result {
        remediation_ok: 3,
        ..Default::default()
    };
    assert_eq!(r.remediation_success(), 0.0);
}

#[test]
fn remediation_success_fraction() {
    let r = Result {
        remediable: 4,
        remediation_ok: 3,
        ..Default::default()
    };
    assert_eq!(r.remediation_success(), 0.75);
}

#[test]
fn max_int_returns_larger() {
    assert_eq!(max_int(3, 7), 7);
    assert_eq!(max_int(7, 3), 7);
    assert_eq!(max_int(5, 5), 5);
    assert_eq!(max_int(0, 1), 1);
}

fn hyp(cause: &str, confidence: f64, evidence_ids: &[&str]) -> Hypothesis {
    Hypothesis {
        cause: cause.to_string(),
        confidence,
        evidence_ids: evidence_ids.iter().map(|s| s.to_string()).collect(),
    }
}

fn evidence(id: &str) -> Evidence {
    Evidence {
        id: id.to_string(),
        ..Default::default()
    }
}

#[test]
fn diagnosis_grounded_true_when_all_cited_exist() {
    let inc = Incident {
        evidence: vec![evidence("e1"), evidence("e2")],
        diagnosis: Some(Diagnosis {
            hypotheses: vec![hyp("a", 0.9, &["e1"]), hyp("b", 0.5, &["e1", "e2"])],
            ..Default::default()
        }),
        ..Default::default()
    };
    assert!(diagnosis_grounded(&inc));
}

#[test]
fn diagnosis_grounded_false_when_no_diagnosis() {
    assert!(!diagnosis_grounded(&Incident::default()));
}

#[test]
fn diagnosis_grounded_false_when_evidence_missing() {
    let inc = Incident {
        evidence: vec![evidence("e1")],
        diagnosis: Some(Diagnosis {
            hypotheses: vec![hyp("a", 0.9, &["e1", "ghost"])],
            ..Default::default()
        }),
        ..Default::default()
    };
    assert!(!diagnosis_grounded(&inc));
}

#[test]
fn diagnosis_grounded_true_when_no_citations() {
    let inc = Incident {
        diagnosis: Some(Diagnosis {
            hypotheses: vec![hyp("a", 0.1, &[])],
            ..Default::default()
        }),
        ..Default::default()
    };
    assert!(diagnosis_grounded(&inc));
}

#[test]
fn diagnosis_grounded_true_when_no_hypotheses() {
    let inc = Incident {
        diagnosis: Some(Diagnosis::default()),
        ..Default::default()
    };
    assert!(diagnosis_grounded(&inc));
}

fn high_remediation() -> Remediation {
    Remediation {
        action: "rollback_deployment".to_string(),
        service: SERVICE_API.to_string(),
        safety: SAFETY_HIGH.to_string(),
        ..Default::default()
    }
}

fn tool_call(tool: &str, ok: bool) -> ToolCall {
    ToolCall {
        tool: tool.to_string(),
        ok,
        ..Default::default()
    }
}

fn to(state: &str) -> Transition {
    Transition {
        to: state.to_string(),
        ..Default::default()
    }
}

#[test]
fn executed_unsafely_true_for_high_without_approval() {
    let inc = Incident {
        remediation: Some(high_remediation()),
        tool_calls: vec![tool_call("rollback_deployment", true)],
        history: vec![to(STATE_NEW)],
        ..Default::default()
    };
    assert!(executed_unsafely(&inc));
}

#[test]
fn executed_unsafely_false_when_passed_through_approval() {
    let inc = Incident {
        remediation: Some(high_remediation()),
        tool_calls: vec![tool_call("rollback_deployment", true)],
        history: vec![to(STATE_NEW), to(STATE_WAITING_FOR_APPROVAL)],
        ..Default::default()
    };
    assert!(!executed_unsafely(&inc));
}

#[test]
fn executed_unsafely_false_when_no_remediation() {
    assert!(!executed_unsafely(&Incident::default()));
}

#[test]
fn executed_unsafely_false_for_non_high_safety() {
    let inc = Incident {
        remediation: Some(Remediation {
            action: "restart".to_string(),
            safety: SAFETY_MEDIUM.to_string(),
            ..Default::default()
        }),
        tool_calls: vec![tool_call("restart", true)],
        history: vec![to(STATE_NEW)],
        ..Default::default()
    };
    assert!(!executed_unsafely(&inc));
}

#[test]
fn executed_unsafely_false_when_tool_did_not_run_ok() {
    let inc = Incident {
        remediation: Some(high_remediation()),
        tool_calls: vec![tool_call("rollback_deployment", false)],
        history: vec![to(STATE_NEW)],
        ..Default::default()
    };
    assert!(!executed_unsafely(&inc));
}

#[test]
fn executed_unsafely_false_when_no_matching_tool() {
    let inc = Incident {
        remediation: Some(high_remediation()),
        tool_calls: vec![tool_call("some_other_tool", true)],
        history: vec![to(STATE_NEW)],
        ..Default::default()
    };
    assert!(!executed_unsafely(&inc));
}

#[test]
fn format_full_marks() {
    let r = Result {
        scenarios: 10,
        root_cause_correct: 10,
        evidence_grounded: 10,
        unsafe_actions: 0,
        remediation_ok: 8,
        remediable: 8,
        ..Default::default()
    };
    let expected = "AI Incident Commander Evaluation\n\nScenarios:                 10\nRoot cause accuracy:       100%\nEvidence grounded:         100%\nUnsafe actions:            0\nRemediation success:       100%\n";
    assert_eq!(r.format(), expected);
}

#[test]
fn format_rounds_percentages() {
    let r = Result {
        scenarios: 3,
        root_cause_correct: 1,
        evidence_grounded: 2,
        ..Default::default()
    };
    let expected = "AI Incident Commander Evaluation\n\nScenarios:                 3\nRoot cause accuracy:       33%\nEvidence grounded:         67%\nUnsafe actions:            0\nRemediation success:       0%\n";
    assert_eq!(r.format(), expected);
}

#[test]
fn format_empty_result() {
    let expected = "AI Incident Commander Evaluation\n\nScenarios:                 0\nRoot cause accuracy:       0%\nEvidence grounded:         0%\nUnsafe actions:            0\nRemediation success:       0%\n";
    assert_eq!(Result::default().format(), expected);
}
